### PR TITLE
Expose update delete storage pools cli 2.5

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

Builds on my branch that exposes update/delete via the API (https://github.com/juju/juju/pull/9674).
Exposes the storage-pool update/delete functionality via the command line commands.

## QA steps

Unit tests included.

## Documentation changes

Command document text included .

## Bug reference

https://bugs.launchpad.net/juju/+bug/1812350

### CODE REVIEW

**Comment by Veebers** [comment]
[delete and update storage pools](https://trello.com/c/rlJvRuiS/596-delete-and-update-storage-pools)

File: cmd/juju/storage/pooldelete_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+func (s *PoolDeleteSuite) TestPoolDeleteErrorsManyArgs(c *gc.C) {
+	_, err := s.runPoolDelete(c, []string{"sunshine", "lollypop"})
+	c.Check(err, gc.ErrorMatches, "pool deletion requires storage pool name")
+}
+
+func (s *PoolUpdateSuite) TestPoolDeleteUnsupportedAPIVersion(c *gc.C) {
+	s.mockAPI.APIVersion = 3
+	_, err := s.runPoolUpdate(c, []string{"sunshine"})
+	c.Check(err, gc.ErrorMatches, "pool update requires name and configuration attributes")
```
Comment (wallyworld) [inline cmd/juju/storage/pooldelete_test.go]: Whaaat? Should be the error message about being not supported by this version of juju

File: cmd/juju/storage/pooldelete_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+type mockPoolDeleteAPI struct {
+	APIVersion int
+}
+
+func (s mockPoolDeleteAPI) DeletePool(pname string) error {
+	return nil
```
Comment (wallyworld) [inline cmd/juju/storage/pooldelete_test.go]: this isn't sufficient - the mock needs to record *what* was deleted and the test needs to check for that

File: cmd/juju/storage/poolupdate.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+// PoolUpdateAPI defines the API methods that the storage commands use.
+type PoolUpdateAPI interface {
+	Close() error
+	UpdatePool(name string, attr map[string]interface{}) error
+	BestAPIVersion() int
+}
+
+const poolUpdateCommandDoc = `
+Update configuration attributes for a single existing storage pool.
```
Comment (wallyworld) [inline cmd/juju/storage/poolupdate.go]: ditto about fixing the examples and see also etc

File: cmd/juju/storage/poolupdate.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+`
+
+// NewPoolUpdateCommand returns a command that updates the named storage pools' attributes.
```
Comment (wallyworld) [inline cmd/juju/storage/poolupdate.go]: replaces....

File: cmd/juju/storage/poolupdate.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	PoolCommandBase
+	newAPIFunc  func() (PoolUpdateAPI, error)
+	poolName    string
+	configAttrs map[string]interface{}
+}
+
+// Init implements Command.Init.
+func (c *poolUpdateCommand) Init(args []string) (err error) {
+	if len(args) < 2 {
+		return errors.New("pool update requires name and configuration attributes")
```
Comment (wallyworld) [inline cmd/juju/storage/poolupdate.go]: and type - we should allow for type also

File: api/storage/client.go [hunk 6]
```diff
@@ -90,6 +90,20 @@ func (c *Client) CreatePool(pname, provider string, attrs map[string]interface{}
 	return c.facade.FacadeCall("CreatePool", args, nil)
 }
 
+// DeletePool deletes the named pool
```
Comment (wallyworld) [inline api/storage/client.go]: needs tests

File: api/storage/client.go [hunk 7]
```diff
@@ -90,6 +90,20 @@ func (c *Client) CreatePool(pname, provider string, attrs map[string]interface{}
...
+func (c *Client) DeletePool(pname string) error {
```
Comment (wallyworld) [inline api/storage/client.go]: To be safe, should do api version check here too. Doesn't necessarily need another explicit test

File: cmd/juju/storage/poolupdate.go (hunk 5 above)
Comment (Veebers) [reply cmd/juju/storage/poolupdate.go]: as per convo, type updates will be handled  via the passed config args.

File: api/storage/client.go [hunk 8]
```diff
@@ -82,12 +82,64 @@ func (c *Client) ListPools(providers, names []string) ([]params.StoragePool, err
 
 // CreatePool creates pool with specified parameters.
 func (c *Client) CreatePool(pname, provider string, attrs map[string]interface{}) error {
-	args := params.StoragePool{
-		Name:     pname,
-		Provider: provider,
-		Attrs:    attrs,
+	if c.BestAPIVersion() < 5 {
```
Comment (wallyworld) [inline api/storage/client.go]: Let's add a comment to say that older facade did not support bulk calls

File: apiserver/allfacades.go [hunk 9]
```diff
@@ -267,8 +267,9 @@ func AllFacades() *facade.Registry {
 
 	reg("StatusHistory", 2, statushistory.NewAPI)
 
-	reg("Storage", 3, storage.NewFacadeV3)
-	reg("Storage", 4, storage.NewFacadeV4) // changes Destroy() method signature.
+	reg("Storage", 3, storage.NewStorageAPIV3)
+	reg("Storage", 4, storage.NewStorageAPIV4) // changes Destroy() method signature.
+	reg("Storage", 5, storage.NewStorageAPI)   // Update and Delete storage pools
```
Comment (wallyworld) [inline apiserver/allfacades.go]: and bulk create

File: apiserver/facades/client/storage/storage.go [hunk 10]
```diff
@@ -422,18 +457,35 @@ func (a *APIv3) validateProviderCriteria(providers []string) error {
...
+func (a *StorageAPI) CreatePool(p params.StoragePoolArgs) (params.ErrorResults, error) {
+	results := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(p.Pools)),
+	}
+	for i, pool := range p.Pools {
+		_, err := a.poolManager.Create(
+			pool.Name,
+			storage.ProviderType(pool.Provider),
+			pool.Attrs)
+		if err != nil {
```
Comment (wallyworld) [inline apiserver/facades/client/storage/storage.go]: no need for nil check - common.ServerError(err) returns nill if err is nil

File: cmd/juju/storage/poolcreate.go [hunk 11]
```diff
@@ -18,6 +18,7 @@ import (
 type PoolCreateAPI interface {
 	Close() error
 	CreatePool(pname, ptype string, pconfig map[string]interface{}) error
+	BestAPIVersion() int
```
Comment (wallyworld) [inline cmd/juju/storage/poolcreate.go]: This is the only change I can see to poolcreate.go
Is it needed?

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**